cpp/src_prims/timeSeries/jones_transform.h

```diff
@@ -0,0 +1,440 @@
+
+#pragma unroll
+    for (int i = 0; i < P_VALUE; ++i) {
+      tmp[i] = newParams[modelIndex * P_VALUE + i];
+      myNewParams[i] = tmp[i];
+    }
+
+    for (int j = P_VALUE - 1; j > 0; --j) {
+      DataT a = myNewParams[j];
+
+      for (int k = 0; k < j; ++k) {
+        tmp[k] = (myNewParams[k] + a * myNewParams[j - k - 1]) / (1 - (a * a));
+      }
+
+      for (int iter = 0; iter < j; ++iter) {
+        myNewParams[iter] = tmp[iter];
+      }
+    }
+
+#pragma unroll
+    for (int i = 0; i < P_VALUE; ++i) {
+      newParams[modelIndex * P_VALUE + i] = 2 * myATanh(myNewParams[i]);
+    }
+  }
+}
+
+/**
+ * @brief kernel to perform jones inverse transformation on the moving average params
+ * @tparam DataT: type of the params
+ * @tparam Q_VALUE: q-paramter for the batch of ARIMA(p,d,q) models
+ * @tparam IdxT: type of indexing
+ * @tparam BLOCK_DIM_X: number of threads in block in x dimension
+ * @tparam BLOCK_DIM_Y: number of threads in block in y dimension
+ * @param newParams: pointer to the memory where the new params are to be stored, which is also where the initial mapped input is stored
+ * @param batchSize: number of models in a batch
+ */
+template <typename DataT, int Q_VALUE, typename IdxT, int BLOCK_DIM_X,
+          int BLOCK_DIM_Y>
+__global__ void ma_param_invtransform_kernel(DataT* newParams, IdxT batchSize) {
+  //calculating the index of the model that the coefficients belong to
+  IdxT modelIndex = threadIdx.x + ((IdxT)blockIdx.x * blockDim.x);
+  IdxT k, j;
+
+  DataT tmp[Q_VALUE];
+  DataT myNewParams[Q_VALUE];
+
+  if (modelIndex < batchSize) {
+    //populating the local memory with the global memory
+
+#pragma unroll
+    for (k = 0; k < Q_VALUE; ++k) {
+      tmp[k] = newParams[modelIndex * Q_VALUE + k];
+      myNewParams[k] = tmp[k];
+    }
+
+    for (j = Q_VALUE - 1; j > 0; --j) {
+      DataT b = myNewParams[j];
+
+      for (k = 0; k < j; ++k) {
+        tmp[k] = (myNewParams[k] - b * myNewParams[j - k - 1]) / (1 - (b * b));
+      }
+
+      for (k = 0; k < j; ++k) {
+        myNewParams[k] = tmp[k];
+      }
+    }
+
+#pragma unroll
+    for (k = 0; k < Q_VALUE; ++k) {
+      newParams[modelIndex * Q_VALUE + k] = 2 * myATanh(myNewParams[k]);
+    }
+  }
+}
+
+/**
+ * @brief kernel to perform jones transformation on the autoregressive params
+ * @tparam DataT: type of the params
+ * @tparam P_VALUE: p-paramter for the batch of ARIMA(p,d,q) models
+ * @tparam IdxT: type of indexing
+ * @tparam BLOCK_DIM_X: number of threads in block in x dimension
+ * @tparam BLOCK_DIM_Y: number of threads in block in y dimension
+ * @param newParams: pointer to the memory where the new params are to be stored, which is also where the initial mapped input is stored
+ * @param batchSize: number of models in a batch
+ */
+template <typename DataT, int P_VALUE, typename IdxT, int BLOCK_DIM_X,
+          int BLOCK_DIM_Y>
+__global__ void ar_param_transform_kernel(DataT* newParams, IdxT batchSize) {

```

Comment:
These the transform / invtransform kernels look almost identical across AR & MA. Is it possible to condense these implementations any further?